crates/ruff_python_parser/src/parser/statement.rs

```diff
@@ -2346,84 +2428,238 @@ impl<'src> Parser<'src> {
-        let mut has_seen_default_param = false;
+        // TODO(dhruvmanila): This has the same problem where if there are multiple
+        // kwarg or vararg, the last one will win and the parser will drop the
+        // previous ones. Another thing is the vararg and kwarg uses `Parameter`
+        // which means that the parser cannot recover well from `*args=(1, 2)`.
+        let mut parameters = ast::Parameters {
+            range: TextRange::default(),
+            posonlyargs: vec![],
+            args: vec![],
+            kwonlyargs: vec![],
+            vararg: None,
+            kwarg: None,
+        };
 
-        let start = self.node_start();
+        let mut seen_default_param = false; // `a=10`
+        let mut seen_positional_only_separator = false; // `/`
+        let mut seen_keyword_only_separator = false; // `*`
+        let mut seen_keyword_only_param_after_separator = false;
+
+        // Range of the keyword only separator if it's the last parameter in the list.
+        let mut last_keyword_only_separator_range = None;
 
         self.parse_comma_separated_list(RecoveryContextKind::Parameters(function_kind), |parser| {
-            // Don't allow any parameter after we have seen a vararg `**kwargs`
-            if has_seen_vararg {
-                parser.add_error(
-                    ParseErrorType::ParamFollowsVarKeywordParam,
-                    parser.current_token_range(),
-                );
+            if parameters.kwarg.is_some() {
+                // TODO(dhruvmanila): This fails AST validation in tests because
+                // of the pre-order visit
+                // test_err params_follows_var_keyword_param
+                // def foo(**kwargs, a, /, b=10, *, *args): ...
+                parser.add_error(ParseErrorType::ParamFollowsVarKeywordParam, parser.current_token_range());
             }
 
-            if parser.eat(TokenKind::Star) {
-                has_seen_asterisk = true;
-                if parser.at(TokenKind::Comma) {
-                    has_seen_default_param = false;
-                } else if parser.at_expr() {
-                    let param = parser.parse_parameter(function_kind);
-                    vararg = Some(Box::new(param));
+            match parser.current_token_kind() {
+                TokenKind::Star => {
+                    let star_range = parser.current_token_range();
+                    parser.bump(TokenKind::Star);
+
+                    if parser.at(TokenKind::Name) {
+                        let param = parser.parse_parameter(function_kind, AllowStarAnnotation::Yes);
+                        let param_star_range = param.range().cover(star_range);
+
+                        if parser.at(TokenKind::Equal) {
+                            // test_err params_var_positional_with_default
+                            // def foo(a, *args=(1, 2)): ...
+                            parser.add_error(ParseErrorType::VarParameterWithDefault, parser.current_token_range());
```

Comment:
Nit and unrelated to this PR. You could try to add `#[cold]` to `add_error` and profile if it speeds up the "happy path" where we have no parse errors.